File: sdk/php/tests/e2e/rest/Account/AccountWithdrawTest.php
<?php

namespace Tests\e2e\rest\Account;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Extension\Interceptor\Logging;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\CancelWithdrawalReq;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\GetWithdrawalHistoryOldReq;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\GetWithdrawalHistoryReq;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\GetWithdrawalQuotasReq;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\WithdrawalApi;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\WithdrawalV1Req;
use KuCoin\UniversalSDK\Generate\Account\Withdrawal\WithdrawalV3Req;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use PHPUnit\Framework\TestCase;

class AccountWithdrawTest extends TestCase
{
    /**
     * @var WithdrawalApi $api
     */
    private $api;

    /**
     * @var Serializer $serializer
     */
    private $serializer;

    protected function setUp(): void
    {

        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        // Retrieve API secret information from environment variables
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        // Optional: Retrieve broker secret information from environment variables; applicable for broker API only
        $brokerName = getenv('BROKER_NAME');
        $brokerKey = getenv('BROKER_KEY');
        $brokerPartner = getenv('BROKER_PARTNER');

        // Set specific options, others will fall back to default values
        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setMaxConnections(10)
            ->setInterceptors([new Logging()])
            ->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setBrokerName($brokerName)
            ->setBrokerKey($brokerKey)
            ->setBrokerPartner($brokerPartner)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->build();

        $client = new DefaultClient($clientOption);
        $kucoinRestService = $client->restService();
        $this->api = $kucoinRestService->getAccountService()->getWithdrawalApi();
    }


    /**
     * getWithdrawalQuotas
     * Get Withdrawal Quotas
     * /api/v1/withdrawals/quotas
     */
    public function testGetWithdrawalQuotas()
    {
        $builder = GetWithdrawalQuotasReq::builder();
        $builder->setCurrency("USDT")->setChain("bsc");
        $req = $builder->build();
        $resp = $this->api->getWithdrawalQuotas($req);
        self::assertNotNull($resp->currency);
        self::assertNotNull($resp->limitBTCAmount);
        self::assertNotNull($resp->usedBTCAmount);
        self::assertNotNull($resp->quotaCurrency);
        self::assertNotNull($resp->limitQuotaCurrencyAmount);
        self::assertNotNull($resp->usedQuotaCurrencyAmount);
        self::assertNotNull($resp->remainAmount);
        self::assertNotNull($resp->availableAmount);
        self::assertNotNull($resp->withdrawMinFee);
        self::assertNotNull($resp->innerWithdrawMinFee);
        self::assertNotNull($resp->withdrawMinSize);
        self::assertNotNull($resp->isWithdrawEnabled);
        self::assertNotNull($resp->precision);
        self::assertNotNull($resp->chain);
        self::assertNotNull($resp->lockedAmount);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * withdrawalV3
     * Withdraw (V3)
     * /api/v3/withdrawals
     */
    public function testWithdrawalV3()
    {
        $builder = WithdrawalV3Req::builder();
        $builder->setCurrency("USDT")->setChain("bsc")->setAmount("20")->setIsInner(false)->setRemark("***")
            ->setFeeDeductType("INTERNAL")->setToAddress("******")->setWithdrawType("ADDRESS");
        $req = $builder->build();
        $resp = $this->api->withdrawalV3($req);
        self::assertNotNull($resp->withdrawalId);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * cancelWithdrawal
     * Cancel Withdrawal
     * /api/v1/withdrawals/{withdrawalId}
     */
    public function testCancelWithdrawal()
    {
        $builder = CancelWithdrawalReq::builder();
        $builder->setWithdrawalId("682411bd537806000718515a");
        $req = $builder->build();
        $resp = $this->api->cancelWithdrawal($req);
        self::assertNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getWithdrawalHistory
     * Get Withdrawal History
     * /api/v1/withdrawals
     */
    public function testGetWithdrawalHistory()
    {
        $builder = GetWithdrawalHistoryReq::builder();
        $builder->setCurrency("USDT");
        $req = $builder->build();
        $resp = $this->api->getWithdrawalHistory($req);
        self::assertNotNull($resp->currentPage);
        self::assertNotNull($resp->pageSize);
        self::assertNotNull($resp->totalNum);
        self::assertNotNull($resp->totalPage);
        foreach ($resp->items as $item) {
            self::assertNotNull($item->id);
            self::assertNotNull($item->currency);
            self::assertNotNull($item->chain);
            self::assertNotNull($item->status);
            self::assertNotNull($item->address);
            self::assertNotNull($item->memo);
            self::assertNotNull($item->isInner);
            self::assertNotNull($item->amount);
            self::assertNotNull($item->fee);
            self::assertNotNull($item->createdAt);
            self::assertNotNull($item->updatedAt);
            self::assertNotNull($item->remark);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getWithdrawalHistoryOld
     * Get Withdrawal History - Old
     * /api/v1/hist-withdrawals
     */
    public function testGetWithdrawalHistoryOld()
    {
        $builder = GetWithdrawalHistoryOldReq::builder();
        $builder->setCurrency("USDT");
        $req = $builder->build();
        $resp = $this->api->getWithdrawalHistoryOld($req);
        self::assertNotNull($resp->currentPage);
        self::assertNotNull($resp->pageSize);
        self::assertNotNull($resp->totalNum);
        self::assertNotNull($resp->totalPage);
        foreach ($resp->items as $item) {
            self::assertNotNull($item->currency);
            self::assertNotNull($item->createAt);
            self::assertNotNull($item->amount);
            self::assertNotNull($item->address);
            self::assertNotNull($item->walletTxId);
            self::assertNotNull($item->isInner);
            self::assertNotNull($item->status);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * withdrawalV1
     * Withdraw - V1
     * /api/v1/withdrawals
     */
    public function testWithdrawalV1()
    {
        $builder = WithdrawalV1Req::builder();
        $builder->setCurrency("USDT")->setChain("bsc")->setAmount("20")->setIsInner(false)->setRemark("***")
            ->setFeeDeductType("INTERNAL")->setAddress("****");

        $req = $builder->build();
        $resp = $this->api->withdrawalV1($req);
        self::assertNotNull($resp->withdrawalId);
        Logger::info($resp->jsonSerialize($this->serializer));
    }


}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/e2e/rest/account/AccountWithdrawTest.java
package com.kucoin.universal.sdk.test.e2e.rest.account;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.account.subaccount.*;
import com.kucoin.universal.sdk.generate.account.withdrawal.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.io.IOException;
import java.util.Collections;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

@Slf4j
public class AccountWithdrawTest {

  private static WithdrawalApi api;

  public static ObjectMapper mapper = new ObjectMapper();

  @BeforeAll
  public static void setUp() {

    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    TransportOption httpTransport =
        TransportOption.builder()
            .interceptors(
                Collections.singleton(
                    new Interceptor() {
                      @NotNull @Override
                      public Response intercept(@NotNull Chain chain) throws IOException {
                        Request request = chain.request();

                        System.out.println("========== Request ==========");
                        System.out.println(request.method() + " " + request.url());

                        Response response = chain.proceed(request);

                        System.out.println("========== Response ==========");
                        System.out.println("Status Code: " + response.code());
                        System.out.println("Message: " + response.message());
                        return response;
                      }
                    }))
            .build();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransport)
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    api = kucoinClient.getRestService().getAccountService().getWithdrawalApi();
  }

  /** getWithdrawalQuotas Get Withdrawal Quotas /api/v1/withdrawals/quotas */
  @Test
  public void testGetWithdrawalQuotas() throws Exception {
    GetWithdrawalQuotasReq.GetWithdrawalQuotasReqBuilder builder = GetWithdrawalQuotasReq.builder();
    builder.currency("USDT").chain("bsc");
    GetWithdrawalQuotasReq req = builder.build();
    GetWithdrawalQuotasResp resp = api.getWithdrawalQuotas(req);
    Assertions.assertNotNull(resp.getCurrency());
    Assertions.assertNotNull(resp.getLimitBTCAmount());
    Assertions.assertNotNull(resp.getUsedBTCAmount());
    Assertions.assertNotNull(resp.getQuotaCurrency());
    Assertions.assertNotNull(resp.getLimitQuotaCurrencyAmount());
    Assertions.assertNotNull(resp.getUsedQuotaCurrencyAmount());
    Assertions.assertNotNull(resp.getRemainAmount());
    Assertions.assertNotNull(resp.getAvailableAmount());
    Assertions.assertNotNull(resp.getWithdrawMinFee());
    Assertions.assertNotNull(resp.getInnerWithdrawMinFee());
    Assertions.assertNotNull(resp.getWithdrawMinSize());
    Assertions.assertNotNull(resp.getIsWithdrawEnabled());
    Assertions.assertNotNull(resp.getPrecision());
    Assertions.assertNotNull(resp.getChain());
    Assertions.assertNotNull(resp.getLockedAmount());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** withdrawalV3 Withdraw (V3) /api/v3/withdrawals */
  @Test
  public void testWithdrawalV3() throws Exception {
    WithdrawalV3Req.WithdrawalV3ReqBuilder builder = WithdrawalV3Req.builder();
    builder
        .currency("USDT")
        .chain("bsc")
        .amount("10")
        .memo("")
        .isInner(false)
        .remark("****")
        .feeDeductType("INTERNAL")
        .toAddress("***")
        .withdrawType(WithdrawalV3Req.WithdrawTypeEnum.ADDRESS);
    WithdrawalV3Req req = builder.build();
    WithdrawalV3Resp resp = api.withdrawalV3(req);
    Assertions.assertNotNull(resp.getWithdrawalId());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** cancelWithdrawal Cancel Withdrawal /api/v1/withdrawals/{withdrawalId} */
  @Test
  public void testCancelWithdrawal() throws Exception {
    CancelWithdrawalReq.CancelWithdrawalReqBuilder builder = CancelWithdrawalReq.builder();
    builder.withdrawalId("68804a886538a800078cdde3");
    CancelWithdrawalReq req = builder.build();
    CancelWithdrawalResp resp = api.cancelWithdrawal(req);
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getWithdrawalHistory Get Withdrawal History /api/v1/withdrawals */
  @Test
  public void testGetWithdrawalHistory() throws Exception {
    GetWithdrawalHistoryReq.GetWithdrawalHistoryReqBuilder builder =
        GetWithdrawalHistoryReq.builder();
    builder.currency("USDT");
    GetWithdrawalHistoryReq req = builder.build();
    GetWithdrawalHistoryResp resp = api.getWithdrawalHistory(req);
    Assertions.assertNotNull(resp.getCurrentPage());
    Assertions.assertNotNull(resp.getPageSize());
    Assertions.assertNotNull(resp.getTotalNum());
    Assertions.assertNotNull(resp.getTotalPage());
    resp.getItems()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getId());
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getChain());
              Assertions.assertNotNull(item.getStatus());
              Assertions.assertNotNull(item.getAddress());
              Assertions.assertNotNull(item.getMemo());
              Assertions.assertNotNull(item.getIsInner());
              Assertions.assertNotNull(item.getAmount());
              Assertions.assertNotNull(item.getFee());
              Assertions.assertNotNull(item.getCreatedAt());
              Assertions.assertNotNull(item.getUpdatedAt());
              Assertions.assertNotNull(item.getRemark());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getWithdrawalHistoryById Get Withdrawal History By ID /api/v1/withdrawals/{withdrawalId} */
  @Test
  public void testGetWithdrawalHistoryById() throws Exception {
    GetWithdrawalHistoryByIdReq.GetWithdrawalHistoryByIdReqBuilder builder =
        GetWithdrawalHistoryByIdReq.builder();
    builder.withdrawalId("68804a886538a800078cdde3");
    GetWithdrawalHistoryByIdReq req = builder.build();
    GetWithdrawalHistoryByIdResp resp = api.getWithdrawalHistoryById(req);
    Assertions.assertNotNull(resp.getId());
    Assertions.assertNotNull(resp.getUid());
    Assertions.assertNotNull(resp.getCurrency());
    Assertions.assertNotNull(resp.getChainId());
    Assertions.assertNotNull(resp.getChainName());
    Assertions.assertNotNull(resp.getCurrencyName());
    Assertions.assertNotNull(resp.getStatus());
    Assertions.assertNotNull(resp.getFailureReason());
    Assertions.assertNotNull(resp.getAddress());
    Assertions.assertNotNull(resp.getMemo());
    Assertions.assertNotNull(resp.getIsInner());
    Assertions.assertNotNull(resp.getAmount());
    Assertions.assertNotNull(resp.getFee());
    Assertions.assertNotNull(resp.getWalletTxId());
    Assertions.assertNotNull(resp.getAddressRemark());
    Assertions.assertNotNull(resp.getRemark());
    Assertions.assertNotNull(resp.getCreatedAt());
    Assertions.assertNotNull(resp.getCancelType());

    Assertions.assertNotNull(resp.getReturnStatus());
    Assertions.assertNotNull(resp.getReturnCurrency());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getWithdrawalHistoryOld Get Withdrawal History - Old /api/v1/hist-withdrawals */
  @Test
  public void testGetWithdrawalHistoryOld() throws Exception {
    GetWithdrawalHistoryOldReq.GetWithdrawalHistoryOldReqBuilder builder =
        GetWithdrawalHistoryOldReq.builder();
    builder.currency("USDT").startAt(1703001600000L).endAt(1703260800000L);
    GetWithdrawalHistoryOldReq req = builder.build();
    GetWithdrawalHistoryOldResp resp = api.getWithdrawalHistoryOld(req);
    Assertions.assertNotNull(resp.getCurrentPage());
    Assertions.assertNotNull(resp.getPageSize());
    Assertions.assertNotNull(resp.getTotalNum());
    Assertions.assertNotNull(resp.getTotalPage());
    resp.getItems()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getCreateAt());
              Assertions.assertNotNull(item.getAmount());
              Assertions.assertNotNull(item.getAddress());
              Assertions.assertNotNull(item.getWalletTxId());
              Assertions.assertNotNull(item.getIsInner());
              Assertions.assertNotNull(item.getStatus());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** withdrawalV1 Withdraw - V1 /api/v1/withdrawals */
  @Test
  public void testWithdrawalV1() throws Exception {
    WithdrawalV1Req.WithdrawalV1ReqBuilder builder = WithdrawalV1Req.builder();
    builder.currency("USDT").chain("bsc").address("***").amount(10L).memo("").isInner(false);
    WithdrawalV1Req req = builder.build();
    WithdrawalV1Resp resp = api.withdrawalV1(req);
    Assertions.assertNotNull(resp.getWithdrawalId());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }
}
